### ISSUE TITLE
"undefined reference" error in `test_piracy`

### ISSUE THREAD

**Issue by dhanak** [JuliaTesting/Aqua.jl#167]
Version 0.6.4 introduced an issue with `test_piracy`, which can be reproduced with the following minimal working example:

```
julia> module M
           struct S <: Function
               i::Int
           end
       end

julia> using Aqua; Aqua.test_piracy(M)
ERROR: UndefRefError: access to undefined reference
```

The problem is that even though `S` is a `Function`, `S.instance` doesn't exist. I guess the latter makes sense, since S can only be instantiated with a specific integer parameter, it doesn't have a “default” instance.

**Comment by lgoettgens** [comment]
Note that this is an issue only on the 0.6.x versions, on master/0.7-DEV this is already fixed. A small fix for 0.6.x is #170.

Furthermore, please note that there is no support to use your own subtypes of `Function` in `treat_as_own`. If you would like to have that feature, this should be added as a new issue. Optimally, this should then include some conceptual idea to implement it or at least the input format for `treat_as_own`.

**Comment by fingolfin** [comment]
@dhanak can you confirm this works in 0.7?
